Update complex interface

jdemel opened the pull request:
This is a draft.

**API break**: This work breaks the API on purpose.

Mostly, it is a try to fix these issues with C vs. C++ complex types. We rely on undefined behavior. This is supposed to end and be better.

Apparently, we hit a bunch of issues:
- `complex` is considered a "niche"
- `complex.h` and `std::complex` seem to be explicitly binary incompatible. At least they're allowed to.
- Issues multiply as soon as we consider non float types. e.g. `char _Complex`. 
- Probably more issues we haven't even encountered yet.

**EDIT**
Fixes #442 
Fixes #572 

Current state:
- C Interface is updated.
- C++ must use C interface and wrap around it.
- C++ usage is somehow less convenient at the moment. Wrappers missing.
- PowerPC issues are fixed.
- CI passes on Linux x86 and arm.
- CI fails on Windows and MacOS.

**EDIT**
Since we want to support compilers that are difficult to support with complex C (i.e. complex C support is missing), we ended up with the current API that basically relies on Undefined Behavior (UB).

jdemel committed:
api: Start to fix complex types

This is a first try at moving to a sane, defined complex API. Before we
were basically relying on undefined behavior. This bites us with
PowerPC.

The current changes work for C but C++ code does not compile yet.

Signed-off-by: Johannes Demel <demel@uni-bremen.de>

jdemel committed:
c++: Make things compile

Signed-off-by: Johannes Demel <demel@uni-bremen.de>

jdemel committed:
wip: Add example files

Signed-off-by: Johannes Demel <demel@uni-bremen.de>

marcusmueller commented:
you can't fix this without an API break, so break API in a visible, and useful manner, and completely remove the same-name C++ wrapper; instead of 

```c++
void volk_32fc_x2_dot_prod_32fc(lv_32fc_t*, const lv_32fc_t*, const lv_32fc_t*, unsigned int)
```

a C++ consumer should be calling

```c++
volk::dot_prod(const std::vector<std::complex<float>>&,const std::vector<std::complex<float>>&, std::vector<std::complex<float>>&)
```

or

```c++
void volk::dot_prod(const std::complex<float>*, const std::complex<float>*, std::complex<float>*, unsigned int)
```

we should be generating these wrappers automatically. 

Ideally, these wrappers wouldn't just be C++ functions, but `static operator(...)` members of a class (per kernel), so that we can later add functionality like querying kernels what their favorite granularity is, which protokernels exist and so on.

jdemel commented:
Thanks for your suggestions. I want to update the interface and that'll break the current API.

The C++ interface should require `volk::vector` because we can assume they're aligned. If we'd use any kind of vector, the C++ interface doesn't need a separate vector length parameter anymore.

However, at the moment, I'm stuck with a more trivial problem. For any interface, we'd need to be able to call a C kernel with an `lv_32fc_t` datatype. Right now, I'm unable to compile any C++ as long as I try to use a C typedef. The issue is: How do I convince my C++ compiler to pick up C compelx data type definitions?

jdemel committed:
complex: Add example interface for C++

Signed-off-by: Johannes Demel <demel@uni-bremen.de>

marcusmueller commented on include/volk/volk_complex.h:
```diff
@@ -17,55 +17,32 @@
-typedef float complex lv_32fc_t;
-typedef double complex lv_64fc_t;
+// Obviously, we would love `typedef float complex lv_32fc_t` to work.
+// However, this clashes with C++ definitions.
```
C++-land should **never** even need to see the C typedef – there's simply no reason. I'll be honest, your `lv_something` typedefs look fun, but what problem are they actually solving? Why is it worth solving at this effort and aesthetic expense?

marcusmueller commented on main.cc:
```diff
@@ -0,0 +1,82 @@
+#include <iostream>
+#include <vector>
+
+typedef std::complex<float> cmplxf;
```
This is a sensible typedef for C++ – no need for having `lv_32fc_t` in C++ land, right?

marcusmueller commented on main.cc:
```diff
@@ -0,0 +1,82 @@
+
+
+void cppmultiply(volk::vector<cmplxf>& result,
+                 volk::vector<cmplxf>& input0,
```
you should `const` these inputs ;)

jdemel committed:
qa: Update QA code to new interface

With this commit, our QA code should work again. However, it still
relies on C types in the C++ domain. We probably want to add a C++
wrapper around C VOLK.

The idea:
- add `volk.hh`
- Include C++ magic in here.
- Find idiomatic, modern C++ interface defition.

Signed-off-by: Johannes Demel <demel@uni-bremen.de>

jdemel committed:
ci: Fix formatting in kernel_tests.h

Signed-off-by: Johannes Demel <demel@uni-bremen.de>

jdemel committed:
ci: Fix formatting in main.c (temp file)

Signed-off-by: Johannes Demel <demel@uni-bremen.de>

jdemel committed:
ci: Add ppc64le and s390x tests to TravisCI

We removed these tests because of our broken interface. Now, we aim to
fix this issue and thus, expect it to work again.

Signed-off-by: Johannes Demel <demel@uni-bremen.de>

jdemel committed:
api: Develop C++ wrapper API

The VOLK C API should be in place.
We stick with the current API. Except, we always require it to be C. No
more C and C++ mix and match.

This allows us to be more open about the C++ API. Here, we write
wrappers around our C functions.

Signed-off-by: Johannes Demel <demel@uni-bremen.de>

jdemel commented:
At the moment I assume, the updated C API is in good shape.
The MSVC error messages in our CI are currently not helpful. This needs further investigation.
The MacOS error is probably due to some `__VOLK_DECL_BEGIN` specifics. We probably want to drop the `&& (__GNUC__)` condition.

The C++ API update is in a more experimental state. It works with 2 examples.

C++ Interface requirements
1. Make C++ STL types usable `std::vector`, `std::complex`.
2. Make aligned vectors aka `volk::vector` usable.
3. Allow call-by-pointer for GR buffer interface usage etc.

These requirements result in at least 3 functions.
We might want to think about fancy new C++ features e.g. concepts to consolidate these.

More considerations: Do we want to expose more info?
e.g.
- explicit function pointers.
  - Especially: Make these explicitly callable?
- adopt a `volk::dot_prod::operator` like interface.
- add option to obtain available implementations for a kernel in C++?
- Add an explicit `volk.hh`, or `volkpp.h`, or `volkcpp.h` header file for the C++ API. That'd include `volk_alloc.hh`.

jdemel committed:
interface: Declare extern C lib

Now we declare extern C for all compilers and hope this works.

Signed-off-by: Johannes Demel <demel@uni-bremen.de>

jdemel committed:
interface: Declare more extern C

Make sure everything that is supposed have C-linkage does have
C-linkage.

Signed-off-by: Johannes Demel <demel@uni-bremen.de>

jdemel commented:
At this point, everything compiles and and all tests pass on most systems with most compilers.
It seems like there are 2 notable exceptions: AppleClang and MSVC. In both cases, I need help.
@michaelld could you have a look at the MacOS issue?
@ryanvolz would you be able to help out? Or do you know whom we could ask for help?

michaelld commented on include/volk/volk_complex.h:
```diff
@@ -17,55 +17,32 @@
-{
-    return x.real();
-}
+// #include <volk/volk_common.h>
```
why not use this & then in the next (valid) line do `__VOLK_DECL_BEGIN` (and also `__VOLK_DECL_END` at the end)?

jdemel replied:
It didn't work with `__VOLK_DECL_BEGIN` etc. previously. However, it seems like it does now.

michaelld commented on main.c:
```diff
@@ -0,0 +1,98 @@
+    //   for (int i = 0; i < 8; i++) {
+    //     printf("%hhi\n", values[i]);
+    //   }
+}
```
endline needed

michaelld commented on main.cc:
```diff
@@ -0,0 +1,200 @@
+        float i = __imag__ val;
+        fmt::print("sizeof(val)={}, {:+.1f}{:+.1f}j\n", sizeof(val), r, i);
+    }
+}
```
endline needed

michaelld commented on compile.sh:
```diff
@@ -0,0 +1,3 @@
+gcc -std=c17 -I/home/johannes/src/volk/include -I/home/johannes/src/volk/build/include -L/home/johannes/src/volk/build/lib -x c main.c -o mainvolkgnuc -lm -lvolk
```
Guessing you didn't mean to include this file ...

michaelld replied:
ah ... guessing you're using this & `main.c` and `main.cc` as part of the draft for testing purposes, and you'll remove these once the PR is out of draft ... yes?

jdemel replied:
Yes. This is a temporary file I intend to remove as soon as I figured out all the details of how to update the interface.

michaelld commented on include/volk/volk_common.h:
```diff
@@ -76,8 +76,9 @@
 ////////////////////////////////////////////////////////////////////////
 // C-linkage declaration macros
 // FIXME: due to the usage of complex.h, require gcc for c-linkage
+// Hope and p*** that extern "C" works for all relevant compilers nowadays.
```
remove `and p***` please ... "Hope" is sufficient ;)

jdemel replied:
Well. It didn't help. but yeah.

michaelld reviewed (commented):
Good starting point! Let me do some macOS testing to see if this builds for me or not & then I can provide more commentary.

michaelld commented:
@jdemel the basic concept of the `extern "C" {` (`__VOLK_DECL_BEGIN`) is to use it only on local code, not on `#include <>`s. Thus in `include/volk/volk_complex.h`, change the start to:
```
#include <volk/volk_common.h>
#include <complex.h>

__VOLK_DECL_BEGIN
```
and then end with
```
#endif /* __GNUC__ */

__VOLK_DECL_END

#endif /* INCLUDE_VOLK_COMPLEX_H */
```

And for `tmpl/volk_typedefs.tmpl.h`:
```
#include <inttypes.h>
#include <volk/volk_common.h>
#include <volk/volk_complex.h>

__VOLK_DECL_BEGIN

%for kern in kernels:
typedef void (*${kern.pname})(${kern.arglist_types});
%endfor

__VOLK_DECL_END
```

with these changes, this PR builds on macOS for me. These changes should be broadly correct for modern compilers.

jdemel committed:
interface: Add review feedback

Use `__VOLK_DECL_BEGIN` etc. again. It seems to work now.

Signed-off-by: Johannes Demel <demel@uni-bremen.de>

jdemel committed:
interface: Add feedback for temp files

`main.c`, `main.cc` and `compile.sh` are temporary files that are
supposed to be removed, once the new interface is stabilized. In order
to remove distractions, these files should follow best practices.

Signed-off-by: Johannes Demel <demel@uni-bremen.de>

jdemel committed:
interface: Incorporate feedback for MacOS

Michael's comments to fix MacOS builds are included now. It compiles on
my system too. Hopefully CI etc. passes too.

Signed-off-by: Johannes Demel <demel@uni-bremen.de>

jdemel commented:
Thanks @michaelld MacOS passes CI now too. 

What are your thoughts for the C++ interface?

michaelld commented:
> Thanks @michaelld MacOS passes CI now too.

Yay! I'll look at the Ubuntu failures too & see if I can work out what's going on there.

> What are your thoughts for the C++ interface?

I like it, in general. I think it's important for us to offer this interface, since there is a lot of complex math that can be accelerated by appropriately-designed kernels.

michaelld commented:
Ah ,... the Ubuntu ones are still working. Looking positive right now, doing `make test`. Not sure if the Windows ones are going or dead; I can't really help there anyway ...

ryanvolz commented:
The Windows failure seems consistent with https://docs.microsoft.com/en-us/cpp/c-runtime-library/complex-math-support?view=msvc-160. Namely, that `_Complex` is not defined with MSVC. You could do some ifdefs and use `_FComplex` and friends instead, but that looks like it will only work for float and double. I'm don't have a lead on what to do for the rest of it.

jdemel commented:
My first idea was to browse through: [MSVC STL headers](https://github.com/microsoft/STL)
I found a reference to `complex.h` but not the C header file `complex.h`. If you happen to know a place where we can find it, that'd be very useful. MSVC has a generic `std::complex` class that might be able to do `std::complex<int>` etc.

Besides, reading through these MSVC specifics, they basically, once again, state: MSVC supports C89-only and only select later extensions. Thus, requiring C11 excludes MSVC.

If we can't fix it, we might be forced to revert to define our own custom structure types. Obviously, that'd be a very unpleasant prospect.

ryanvolz commented:
> If you happen to know a place where we can find it, that'd be very useful.

It's distributed as part of the Universal C Runtime, which comes in the [Windows SDK](https://developer.microsoft.com/en-us/windows/downloads/windows-10-sdk/). I haven't found a way of viewing just the header without installing the whole thing on a Windows machine, but that's a place to start.

jdemel commented:
@ryanvolz  Thanks for the hints where to search!
From MSVC `complex.h`
```c
typedef struct _C_float_complex
{
    float _Val[2];
} _C_float_complex;

typedef _C_float_complex   _Fcomplex;
```
Thus, we should be able to do things like:
```c
typedef _Fcomplex lv_32fc_t;
typedef _Dcomplex lv_64fc_t;
```
However, we might be forced to define these structs manually. 
```c
typedef char _Complex lv_8sc_t;
typedef short _Complex lv_16sc_t;
typedef long _Complex lv_32sc_t;
typedef long long _Complex lv_64sc_t;
```
In that case, we might be forced to overload complex math functions as well. Thus, whenever one adds a function that uses these types and some complex math functions, we might need to implement these for a new type. However, I assume we'd just do the convert to `double` and back again thing. It is probably inefficient but should only be required for `generic` kernels.

jdemel committed:
interface: Start MSVC fixes

MSVC does not support C complex numbers.
There's a different approach here. Thus, we try to fix it with some
additional burden. Let's poke the CI.

Signed-off-by: Johannes Demel <demel@uni-bremen.de>

jdemel committed:
msvc: Check for _MSC_VER

MSVC requires special treatment. We check for it now.

Signed-off-by: Johannes Demel <demel@uni-bremen.de>

jdemel commented:
I tried to define
```c
typedef _Fcomplex lv_32fc_t;
```
for MSVC, but the CI tells me
```
error C4430: missing type specifier - int assumed. Note: C++ does not support default-int
```
There's something wrong. At this point, I need help from someone with such an environment.

jdemel committed:
complex: Add more references

It is difficult to get MSVC to compile with complex numbers. Add more
referenceses to sources that explain the reasons.

Signed-off-by: Johannes Demel <demel@uni-bremen.de>

ryanvolz commented:
> Currently, my idea would be: Use Clang to compile VOLK with MSVC ABI compat and use MSVC to compile tests, volk_profile etc.

From my conda perspective, using Clang to compile VOLK shouldn't be a problem, although I haven't tried it and I am not aware of the full implications of doing that for downstream consumers (i.e. GNU Radio).

jdemel commented:
I don't have any experience with that approach.
I assume, if we find a way to compile VOLK with Clang on Windows and it is possible to build GNU Radio against it we're good to go.

jdemel commented:
With VOLK 3.1 we went the temporary route to pass complex values by pointer exclusively. While this is not the nicest solution, it works.

jdemel commented:
Essentially, we went with pointers for s32fc. While this is an interesting approach, I close it.
